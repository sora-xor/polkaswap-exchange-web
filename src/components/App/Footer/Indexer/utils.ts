import { IndexerType } from '@soramitsu/soraneo-wallet-web/lib/consts';

export function getIndexerName(type: IndexerType): string {
  return type.charAt(0).toUpperCase() + type.slice(1).toLowerCase();
}
