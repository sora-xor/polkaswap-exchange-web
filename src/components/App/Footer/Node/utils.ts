import { countryCodeEmoji } from 'country-code-emoji';

import store from '@/store';

export function formatLocation(code: string): Nullable<{ flag: string; name?: string }> {
  try {
    const isoCode = code.toUpperCase();
    const flag = countryCodeEmoji(isoCode);
    const location = { flag, name: '' };
    const displayRegions = store.state.settings.displayRegions;
    if (!displayRegions) {
      return location;
    }
    const name = displayRegions.of(isoCode);
    if (name) {
      location.name = name;
    }
    return location;
  } catch (error) {
    console.warn('Unsupported format of location in env.json', error);
    return null;
  }
}
