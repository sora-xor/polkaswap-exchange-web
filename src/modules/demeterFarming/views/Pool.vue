<template>
  <div>
    <pool-base v-bind="{ parentLoading, ...$attrs }" v-on="$listeners">
      <template #title-append="{ liquidity, activeCollapseItems }">
        <div v-if="getStatusBadgeVisibility(liquidity.address, activeCollapseItems)" class="s-flex farming-pool-badges">
          <status-badge
            v-for="item in getAvailablePools(pools[liquidity.secondAddress])"
            :key="`${item.pool.poolAsset}-${item.pool.rewardAsset}`"
            :liquidity="liquidity"
            :pool="item.pool"
            :account-pool="item.accountPool"
            @add="changePoolStake($event, true)"
            class="farming-pool-badge"
          />
        </div>
      </template>
      <template #append="liquidity">
        <pool-card
          v-for="item in getAvailablePools(pools[liquidity.secondAddress])"
          :key="`${item.pool.poolAsset}-${item.pool.rewardAsset}`"
          :liquidity="liquidity"
          :pool="item.pool"
          :account-pool="item.accountPool"
          @add="changePoolStake($event, true)"
          @remove="changePoolStake($event, false)"
          @claim="claimPoolRewards"
          @calculator="showPoolCalculator"
          border
          class="demeter-pool"
        />
      </template>
    </pool-base>

    <stake-dialog
      :visible.sync="showStakeDialog"
      :liquidity="selectedAccountLiquidity"
      :pool="selectedPool"
      :account-pool="selectedAccountPool"
      :is-adding="isAddingStake"
      @add="handleStakeAction($event, deposit)"
      @remove="handleStakeAction($event, withdraw)"
    />
    <claim-dialog
      :visible.sync="showClaimDialog"
      :pool="selectedPool"
      :account-pool="selectedAccountPool"
      @confirm="handleClaimRewards"
    />

    <calculator-dialog
      :visible.sync="showCalculatorDialog"
      :pool="selectedPool"
      :account-pool="selectedAccountPool"
      :liquidity="selectedAccountLiquidity"
    />
  </div>
</template>

<script lang="ts">
import { Component, Mixins } from 'vue-property-decorator';
import { mixins } from '@soramitsu/soraneo-wallet-web';

import PageMixin from '../mixins/PageMixin';

import { demeterLazyComponent } from '../router';
import { DemeterComponents } from '../consts';

import { lazyView } from '@/router';
import { PageNames } from '@/consts';

import { state } from '@/store/decorators';

import type { AccountLiquidity } from '@sora-substrate/util/build/poolXyk/types';

@Component({
  inheritAttrs: false,
  components: {
    PoolBase: lazyView(PageNames.Pool),
    PoolCard: demeterLazyComponent(DemeterComponents.PoolCard),
    StatusBadge: demeterLazyComponent(DemeterComponents.StatusBadge),
    StakeDialog: demeterLazyComponent(DemeterComponents.StakeDialog),
    ClaimDialog: demeterLazyComponent(DemeterComponents.ClaimDialog),
    CalculatorDialog: demeterLazyComponent(DemeterComponents.CalculatorDialog),
  },
})
export default class DemeterPools extends Mixins(PageMixin, mixins.TransactionMixin) {
  @state.pool.accountLiquidity private accountLiquidity!: Array<AccountLiquidity>;

  get selectedAccountLiquidity(): Nullable<AccountLiquidity> {
    return this.accountLiquidity.find((liquidity) => liquidity.secondAddress === this.poolAsset) ?? null;
  }
}
</script>

<style lang="scss" scoped>
.farming-pool-badges {
  flex-flow: wrap;
}
.demeter-pool {
  margin-top: $inner-spacing-medium;
}
</style>
