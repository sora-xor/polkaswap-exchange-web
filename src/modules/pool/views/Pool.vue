<template>
  <div v-loading="parentLoading" class="container el-form--pool">
    <generic-page-header class="page-header--pool" :title="t('exchange.Pool')" :tooltip="t('pool.description')" />
    <div class="pool-wrapper" data-test-name="Pools">
      <p v-if="!isLoggedIn" key="not-logged" class="pool-info-container pool-info-container--empty">
        {{ t('pool.connectToWallet') }}
      </p>
      <p v-else-if="!hasAccountLiquidities" key="pools-empty" class="pool-info-container pool-info-container--empty">
        {{ t('pool.liquidityNotFound') }}
      </p>
      <s-collapse v-else key="has-pools" class="pool-list" :borders="true" @change="updateActiveCollapseItems">
        <s-collapse-item
          v-for="liquidityItem of accountLiquidityData"
          :key="liquidityItem.address"
          :name="liquidityItem.address"
          class="pool-info-container"
        >
          <template #title>
            <pair-token-logo
              :first-token="liquidityItem.firstAsset"
              :second-token="liquidityItem.secondAsset"
              size="small"
            />
            <div class="pool-info-container-block">
              <h3 class="pool-info-container__title">
                {{ liquidityItem.title }}
              </h3>
              <slot name="title-append" v-bind="{ liquidity: liquidityItem, activeCollapseItems }" />
            </div>
          </template>

          <pool-info>
            <info-line
              is-formatted
              value-can-be-hidden
              :label="t('pool.pooledToken', { tokenSymbol: liquidityItem.firstAssetSymbol })"
              :value="liquidityItem.firstBalanceFormatted"
              :fiat-value="liquidityItem.firstBalanceFiat"
            />
            <info-line
              is-formatted
              value-can-be-hidden
              :label="t('pool.pooledToken', { tokenSymbol: liquidityItem.secondAssetSymbol })"
              :value="liquidityItem.secondBalanceFormatted"
              :fiat-value="liquidityItem.secondBalanceFiat"
            />
            <info-line value-can-be-hidden :label="t('pool.poolShare')" :value="liquidityItem.poolShareFormatted" />
            <info-line
              v-if="liquidityItem.apyFormatted"
              :label="t('pool.strategicBonusApy')"
              :value="liquidityItem.apyFormatted"
            />

            <template #buttons>
              <s-button
                type="secondary"
                class="s-typography-button--medium"
                data-test-name="addLiquidity"
                @click="handleAddLiquidity(liquidityItem)"
              >
                {{ t('pool.addLiquidity') }}
              </s-button>
              <s-button
                type="secondary"
                class="s-typography-button--medium"
                data-test-name="removeLiquidity"
                @click="handleRemoveLiquidity(liquidityItem)"
              >
                {{ t('pool.removeLiquidity') }}
              </s-button>
            </template>
          </pool-info>

          <slot name="append" v-bind="{ liquidity: liquidityItem, activeCollapseItems }" />
        </s-collapse-item>
      </s-collapse>
    </div>
    <s-button
      v-if="isLoggedIn"
      class="el-button--add-liquidity s-typography-button--large"
      data-test-name="addLiquidity"
      type="primary"
      @click="handleAddLiquidity()"
    >
      {{ t('pool.addLiquidity') }}
    </s-button>
    <s-button v-else type="primary" class="s-typography-button--large" @click="connectSoraWallet">
      {{ t('connectWalletText') }}
    </s-button>

    <add-liquidity-dialog :visible.sync="addLiquidityVisibility" />
    <remove-liquidity-dialog :visible.sync="removeLiquidityVisibility" />
  </div>
</template>

<script lang="ts">
import { XOR } from '@sora-substrate/sdk/build/assets/consts';
import { mixins, components, WALLET_CONSTS } from '@soramitsu/soraneo-wallet-web';
import { Component, Mixins } from 'vue-property-decorator';

import InternalConnectMixin from '@/components/mixins/InternalConnectMixin';
import { Components } from '@/consts';
import { PoolComponents } from '@/modules/pool/consts';
import PoolApyMixin from '@/modules/pool/mixins/PoolApy';
import { poolLazyComponent } from '@/modules/pool/router';
import { lazyComponent } from '@/router';
import { action, getter, mutation, state } from '@/store/decorators';
import type { LiquidityParams } from '@/store/pool/types';
import { sortPools } from '@/utils';

import type { AccountAsset } from '@sora-substrate/sdk/build/assets/types';
import type { AccountLiquidity } from '@sora-substrate/sdk/build/poolXyk/types';

type LiquidityItem = AccountLiquidity & {
  firstAsset: AccountAsset;
  firstAssetSymbol?: string;
  firstBalanceFormatted?: string;
  firstBalanceFiat?: Nullable<string>;
  secondAsset: AccountAsset;
  secondAssetSymbol?: string;
  secondBalanceFormatted?: string;
  secondBalanceFiat?: Nullable<string>;
  poolShareFormatted?: string;
  apyFormatted?: string;
  title?: string;
};

@Component({
  components: {
    GenericPageHeader: lazyComponent(Components.GenericPageHeader),
    PairTokenLogo: lazyComponent(Components.PairTokenLogo),
    PoolInfo: lazyComponent(Components.PoolInfo),
    AddLiquidityDialog: poolLazyComponent(PoolComponents.AddLiquidityDialog),
    RemoveLiquidityDialog: poolLazyComponent(PoolComponents.RemoveLiquidityDialog),
    FormattedAmount: components.FormattedAmount,
    InfoLine: components.InfoLine,
  },
})
export default class Pool extends Mixins(
  mixins.FormattedAmountMixin,
  mixins.LoadingMixin,
  InternalConnectMixin,
  PoolApyMixin
) {
  readonly FontSizeRate = WALLET_CONSTS.FontSizeRate;
  readonly FontWeightRate = WALLET_CONSTS.FontWeightRate;

  @state.pool.accountLiquidity private accountLiquidity!: Array<AccountLiquidity>;

  @getter.assets.assetDataByAddress private getAsset!: (addr?: string) => Nullable<AccountAsset>;

  @action.addLiquidity.setDataFromLiquidity private setAddressesToAdd!: (args: LiquidityParams) => Promise<void>;

  @mutation.removeLiquidity.setAddresses private setAddressesToRemove!: (args: LiquidityParams) => void;

  activeCollapseItems: string[] = [];

  addLiquidityVisibility = false;
  removeLiquidityVisibility = false;

  get hasAccountLiquidities(): boolean {
    return this.accountLiquidity.length !== 0;
  }

  get accountLiquidityData() {
    const items = this.accountLiquidity.map((liquidity) => {
      const firstAsset = this.getAsset(liquidity.firstAddress) as AccountAsset;
      const firstAssetSymbol = this.getAssetSymbol(firstAsset);
      const secondAsset = this.getAsset(liquidity.secondAddress) as AccountAsset;
      const secondAssetSymbol = this.getAssetSymbol(secondAsset);

      return {
        ...liquidity,
        firstAsset,
        firstAssetSymbol,
        firstBalanceFormatted: this.formatCodecNumber(liquidity.firstBalance, liquidity.decimals),
        firstBalanceFiat: firstAsset ? this.getFiatAmountByCodecString(liquidity.firstBalance, firstAsset) : '0',
        secondAsset,
        secondAssetSymbol,
        secondBalanceFormatted: this.formatCodecNumber(liquidity.secondBalance, liquidity.decimals),
        secondBalanceFiat: secondAsset ? this.getFiatAmountByCodecString(liquidity.secondBalance, secondAsset) : '0',
        poolShareFormatted: `${this.formatStringValue(liquidity.poolShare)}%`,
        apyFormatted: this.getStrategicBonusApy(liquidity.firstAddress, liquidity.secondAddress),
        title: this.getPairTitle(firstAssetSymbol, secondAssetSymbol),
      };
    });

    const defaultSorted = [...items].sort((a, b) =>
      sortPools(
        { baseAsset: a.firstAsset, poolAsset: a.secondAsset },
        { baseAsset: b.firstAsset, poolAsset: b.secondAsset }
      )
    );

    return defaultSorted;
  }

  updateActiveCollapseItems(items: string[]) {
    this.activeCollapseItems = items;
  }

  handleAddLiquidity(item?: LiquidityItem): void {
    const firstAddress = item?.firstAsset.address ?? XOR.address;
    const secondAddress = item?.secondAsset.address ?? '';

    this.setAddressesToAdd({ firstAddress, secondAddress });

    this.addLiquidityVisibility = true;
  }

  handleRemoveLiquidity(item: LiquidityItem): void {
    const firstAddress = item.firstAsset.address;
    const secondAddress = item.secondAsset.address;

    this.setAddressesToRemove({ firstAddress, secondAddress });

    this.removeLiquidityVisibility = true;
  }

  private getAssetSymbol(asset: Nullable<AccountAsset>): string {
    return asset?.symbol ?? this.t('unknownAssetText');
  }

  private getPairTitle(firstTokenSymbol?: string, secondTokenSymbol?: string): string {
    if (firstTokenSymbol && secondTokenSymbol) {
      return `${firstTokenSymbol}-${secondTokenSymbol}`;
    }
    return '';
  }

  private getStrategicBonusApy(firstAddress: string, secondAddress: string): string {
    const apy = this.getPoolApy(firstAddress, secondAddress);
    if (!apy) {
      return '';
    }
    return `${this.getFPNumberFromCodec(apy).mul(this.Hundred).toLocaleString()}%`;
  }
}
</script>

<style lang="scss">
.pool-list {
  @include collapse-items;
  .el-collapse-item__header {
    align-items: flex-start;

    .pair-logo {
      margin-right: $inner-spacing-medium;
      margin-top: $inner-spacing-tiny;
    }
  }
}

@include mobile(true) {
  .pool-info-buttons {
    button {
      font-size: var(--s-font-size-mini);
    }
  }
}
</style>

<style lang="scss" scoped>
$title-height: 42px;

.el-form--pool {
  display: flex;
  flex-direction: column;
  align-items: center;
  .page-header--pool {
    .el-button--settings {
      margin-left: auto;
    }
  }
  .el-button {
    &--create-pair {
      margin-left: 0;
    }
  }
  @include full-width-button;
  @include full-width-button('el-button--create-pair', $inner-spacing-mini);
}

.pool {
  &-wrapper {
    width: 100%;
  }
  &-list {
    width: 100%;
    border-top: none;
    border-bottom: none;
  }
  &-info {
    &-container {
      &--empty {
        color: var(--s-color-base-content-secondary);
        padding: $basic-spacing-medium $inner-spacing-big;
        background: var(--s-color-utility-surface);
        border-radius: var(--s-border-radius-small);
        box-shadow: var(--s-shadow-dialog);
        font-size: var(--s-font-size-small);
        line-height: var(--s-line-height-medium);
        font-weight: 600;
        text-transform: uppercase;
        text-align: center;
      }

      &-block {
        flex: 1;
      }

      &__title {
        flex: 1;
        font-weight: 700;
        text-align: left;
        min-height: $title-height;
        line-height: $title-height;
      }

      & + .el-button {
        margin-top: $inner-spacing-medium;
      }
    }
    & > .asset-logo {
      margin-right: $inner-spacing-mini;
    }
  }
}
</style>
