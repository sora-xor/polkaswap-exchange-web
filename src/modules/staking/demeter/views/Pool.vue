<template>
  <div>
    <pool-base v-bind="{ parentLoading, ...$attrs }" v-on="$listeners">
      <template #title-append="{ liquidity, activeCollapseItems }">
        <div v-show="!isActiveCollapseItem(liquidity.address, activeCollapseItems)" class="s-flex farming-pool-badges">
          <status-badge
            v-for="(item, index) in getLiquidityFarmingPools(liquidity)"
            :key="`${item.pool.poolAsset}-${item.pool.rewardAsset}-${index}`"
            :liquidity="liquidity"
            :pool="item.pool"
            :account-pool="item.accountPool"
            :pool-asset="item.poolAsset"
            :reward-asset="item.rewardAsset"
            :apr="item.apr"
            @add="changePoolStake($event, true)"
            class="farming-pool-badge"
          />
        </div>
      </template>
      <template #append="{ liquidity, activeCollapseItems }">
        <template v-if="isActiveCollapseItem(liquidity.address, activeCollapseItems)">
          <pool-card
            v-for="(item, index) in getLiquidityFarmingPools(liquidity)"
            :key="`${item.pool.poolAsset}-${item.pool.rewardAsset}-${index}`"
            :liquidity="liquidity"
            :pool="item.pool"
            :account-pool="item.accountPool"
            :base-asset="item.baseAsset"
            :pool-asset="item.poolAsset"
            :reward-asset="item.rewardAsset"
            :apr="item.apr"
            :tvl="item.tvl"
            @add="changePoolStake($event, true)"
            @remove="changePoolStake($event, false)"
            @claim="claimPoolRewards"
            @calculator="showPoolCalculator"
            border
            class="demeter-pool"
          />
        </template>
      </template>
    </pool-base>

    <stake-dialog
      :visible.sync="showStakeDialog"
      :is-adding="isAddingStake"
      :liquidity="selectedAccountLiquidity"
      :parent-loading="parentLoading || loading"
      v-bind="selectedDerivedPool"
      @add="handleStakeAction($event, deposit)"
      @remove="handleStakeAction($event, withdraw)"
    />

    <claim-dialog
      :visible.sync="showClaimDialog"
      :parent-loading="parentLoading || loading"
      v-bind="selectedDerivedPool"
      @confirm="handleClaimRewards"
    />

    <calculator-dialog
      :visible.sync="showCalculatorDialog"
      :liquidity="selectedAccountLiquidity"
      v-bind="selectedDerivedPool"
    />
  </div>
</template>

<script lang="ts">
import { mixins } from '@soramitsu/soraneo-wallet-web';
import { Component, Mixins } from 'vue-property-decorator';

import { PoolPageNames } from '@/modules/pool/consts';
import { poolLazyView } from '@/modules/pool/router';
import type { DemeterPoolDerivedData } from '@/modules/staking/demeter/types';
import { state } from '@/store/decorators';

import { demeterStakingLazyComponent } from '../../router';
import { DemeterStakingComponents } from '../consts';
import PageMixin from '../mixins/PageMixin';

import type { AccountLiquidity } from '@sora-substrate/sdk/build/poolXyk/types';

@Component({
  inheritAttrs: false,
  components: {
    PoolBase: poolLazyView(PoolPageNames.Pool),
    PoolCard: demeterStakingLazyComponent(DemeterStakingComponents.PoolCard),
    StatusBadge: demeterStakingLazyComponent(DemeterStakingComponents.StatusBadge),
    StakeDialog: demeterStakingLazyComponent(DemeterStakingComponents.StakeDialog),
    ClaimDialog: demeterStakingLazyComponent(DemeterStakingComponents.ClaimDialog),
    CalculatorDialog: demeterStakingLazyComponent(DemeterStakingComponents.CalculatorDialog),
  },
})
export default class DemeterPools extends Mixins(PageMixin, mixins.TransactionMixin) {
  @state.pool.accountLiquidity private accountLiquidity!: Array<AccountLiquidity>;

  get selectedAccountLiquidity(): Nullable<AccountLiquidity> {
    return (
      this.accountLiquidity.find(
        (liquidity) => liquidity.firstAddress === this.baseAsset && liquidity.secondAddress === this.poolAsset
      ) ?? null
    );
  }

  get selectedDerivedPool(): Nullable<DemeterPoolDerivedData> {
    if (!this.selectedPool) return null;

    return this.prepareDerivedPoolData(this.selectedPool, this.selectedAccountPool, this.selectedAccountLiquidity);
  }

  get farmingPoolsByLiquidities(): Record<string, DemeterPoolDerivedData[]> {
    return this.accountLiquidity.reduce((buffer, liquidity) => {
      const key = this.getLiquidityKey(liquidity);
      const derivedPools = this.getDerivedPools(this.pools[liquidity.firstAddress]?.[liquidity.secondAddress]);

      buffer[key] = derivedPools.map((derived) =>
        this.prepareDerivedPoolData(derived.pool, derived.accountPool, liquidity)
      );

      return buffer;
    }, {});
  }

  getLiquidityFarmingPools(liquidity: AccountLiquidity) {
    return this.farmingPoolsByLiquidities[this.getLiquidityKey(liquidity)] ?? [];
  }

  private getLiquidityKey(liquidity: AccountLiquidity): string {
    return [liquidity.firstAddress, liquidity.secondAddress].join(';');
  }
}
</script>

<style lang="scss" scoped>
.farming-pool-badges {
  flex-flow: wrap;
  gap: $inner-spacing-tiny;
}
.demeter-pool {
  margin-top: $inner-spacing-medium;
}
</style>
