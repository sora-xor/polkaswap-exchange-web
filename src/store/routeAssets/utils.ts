import { FPNumber, Operation } from '@sora-substrate/util/build';
import { Asset, AccountAsset } from '@sora-substrate/util/build/assets/types';
import { api } from '@soramitsu/soraneo-wallet-web';
import { FiatPriceObject } from '@soramitsu/soraneo-wallet-web/lib/services/subquery/types';

import { Recipient } from './types';

import type { WhitelistArrayItem } from '@sora-substrate/util/build/assets/types';

export default {
  validate(recipient: Recipient) {
    return (
      this.wallet(recipient.wallet) &&
      this.asset(recipient.asset) &&
      this.usd(recipient.usd) &&
      this.amount(recipient.amount)
    );
  },

  name(name: Nullable<string>) {
    return name;
  },

  wallet(wallet: string) {
    return api.validateAddress(wallet);
  },

  asset(asset: Asset | AccountAsset) {
    return asset;
  },

  usd(usd: Nullable<number>) {
    return usd && !isNaN(usd);
  },

  amount(amount: Nullable<number>) {
    return amount && !isNaN(amount) && Number.isFinite(amount);
  },
};

export function getTokenEquivalent(
  priceObject: FiatPriceObject,
  asset: Asset | AccountAsset | WhitelistArrayItem,
  usd: number | string
): FPNumber {
  return new FPNumber(usd).div(FPNumber.fromCodecValue(priceObject[asset.address], asset.decimals));
}

export function getAssetUSDPrice(asset, fiatPriceObject) {
  if (!asset) return null;
  return FPNumber.fromCodecValue(fiatPriceObject[asset.address], asset.decimals);
}
