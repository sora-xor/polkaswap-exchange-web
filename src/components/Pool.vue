<template>
  <div class="el-form--pool">
    <!-- TODO: Add appropriate tooltip -->
    <generic-header class="header--pool" :has-button-back="false" :title="t('pool.yourLiquidity')" :tooltip="t('pool.loremIpsum')" />
    <p v-if="!connected" class="pool-info-container">
      {{ t('pool.connectToWallet') }}
    </p>
    <p v-else-if="!liquidities || !liquidities.length" class="pool-info-container">
      {{ t('pool.liquidityNotFound') }}
    </p>
    <s-collapse v-else class="pool-list" :borders="true">
      <s-collapse-item v-for="liquidity of liquidities" :key="liquidity.id" :name="liquidity.id" class="pool-info-container">
        <template #title>
          <pair-token-logo :first-token-symbol="liquidity.firstToken" :second-token-symbol="liquidity.secondToken" size="small" />
          <h3>{{ getPairTitle(liquidity.firstToken, liquidity.secondToken) }}</h3>
        </template>
        <div class="pool-info">
          <token-logo :token-symbol="liquidity.firstToken" size="small" />
          <div>{{ t('pool.pooledToken', { tokenSymbol: liquidity.firstToken }) }}</div>
          <div v-if="liquidity.firstTokenAmount" class="pool-info-value">{{ liquidity.firstTokenAmount }}</div>
        </div>
        <div class="pool-info">
          <token-logo :token-symbol="liquidity.secondToken" size="small" />
          <div>{{ t('pool.pooledToken', { tokenSymbol: liquidity.secondToken }) }}</div>
          <div v-if="liquidity.secondTokenAmount" class="pool-info-value">{{ liquidity.secondTokenAmount }}</div>
        </div>
        <div class="pool-info">
          <pair-token-logo :first-token-symbol="liquidity.firstToken" :second-token-symbol="liquidity.secondToken" size="mini" />
          <div>{{ t('pool.pairTokens', { pair: getPairTitle(liquidity.firstToken, liquidity.secondToken) }) }}</div>
          <div class="pool-info-value">{{ pairValue }}</div>
        </div>
        <div class="pool-info pool-info--share">
          <div>{{ t('pool.poolShare')}}</div>
          <div class="pool-info-value">1.5%</div>
        </div>
        <div class="pool-info--buttons">
          <s-button type="primary" size="small" @click="handleAddPairLiquidity(liquidity.id)">
            {{ t('pool.addLiquidity') }}
          </s-button>
          <s-button type="primary" size="small" @click="handleRemoveLiquidity(liquidity.id)">
            {{ t('pool.removeLiquidity') }}
          </s-button>
        </div>
      </s-collapse-item>
    </s-collapse>
    <s-button class="el-button--add-liquidity" type="primary" @click="handleAddLiquidity">
      {{ t('pool.addLiquidity') }}
    </s-button>
    <s-button v-if="connected" class="el-button--create-pair" type="secondary" @click="handleCreatePair">
      {{ t('pool.createPair') }}
    </s-button>
  </div>
</template>

<script lang="ts">
import { Component, Mixins } from 'vue-property-decorator'
import { Action, Getter } from 'vuex-class'
import TranslationMixin from '@/components/mixins/TranslationMixin'
import { isWalletConnected, formatNumber } from '@/utils'
import router, { lazyComponent } from '@/router'
import { Components, PageNames } from '@/consts'

const namespace = 'pool'

@Component({
  components: {
    GenericHeader: lazyComponent(Components.GenericHeader),
    TokenLogo: lazyComponent(Components.TokenLogo),
    PairTokenLogo: lazyComponent(Components.PairTokenLogo)
  }
})
export default class Pool extends Mixins(TranslationMixin) {
  @Getter('liquidities', { namespace }) liquidities!: any
  @Action('getLiquidities', { namespace }) getLiquidities

  created () {
    this.getLiquidities()
  }

  get connected (): boolean {
    return isWalletConnected()
  }

  get pairValue (): string {
    // TODO: Play with Pair value
    return formatNumber(10.00000171, 4)
  }

  handleAddLiquidity (): void {
    router.push({ name: PageNames.AddLiquidity })
  }

  handleCreatePair (): void {
    router.push({ name: PageNames.CreatePair })
  }

  handleAddPairLiquidity (id): void {
    router.push({ name: PageNames.AddLiquidityId, params: { id } })
  }

  handleRemoveLiquidity (id): void {
    router.push({ name: PageNames.RemoveLiquidity, params: { id } })
  }

  getPairTitle (firstToken, secondToken): string {
    if (firstToken && secondToken) {
      return `${firstToken}-${secondToken}`
    }
    return ''
  }
}
</script>

<style lang="scss">
$pair-icon-height: 36px;
$pool-collapse-icon-height: 2px;
$pool-collapse-icon-width: 10px;

.pool-list {
  .el-collapse-item {
    &__header,
    &__wrap {
      border-bottom: none;
    }
    &__content {
      margin-top: $inner-spacing-medium;
      padding-top: $basic-spacing * 2;
      padding-bottom: 0;
      border-top: 1px solid var(--s-color-base-border-primary);
    }
    .el-collapse-item__header {
      height: $pair-icon-height;
      line-height: $pair-icon-height;
      .pair-logo {
        margin-right: $inner-spacing-medium;
      }
    }
    .el-icon-arrow-right {
      position: relative;
      &:before,
      &:after {
        position: absolute;
        display: block;
        content: '';
        border-radius: 2px;
        background-color: var(--s-color-base-content-primary);
        top: 0;
        right: 0;
        bottom: 0;
        left: 0;
        margin: auto;
      }
      &:before {
        height: $pool-collapse-icon-height;
        width: $pool-collapse-icon-width;
      }
      &:after {
        height: $pool-collapse-icon-width;
        width: $pool-collapse-icon-height;
      }
      &.is-active:after {
        background-color: transparent;
      }
    }
  }
  .el-icon-arrow-right {
    margin-right: 0;
    border-radius: var(--s-border-radius-small);
  }
}
</style>

<style lang="scss" scoped>
$pair-icon-height: 36px;

.header.header--pool {
  margin-top: $inner-spacing-mini;
  margin-bottom: $inner-spacing-mini;
}

.el-form--pool {
  display: flex;
  flex-direction: column;
  align-items: center;
  .el-button {
    &--create-pair {
      margin-left: 0;
      color: var(--s-color-theme-accent);
      &:hover,
      &:active,
      &:disabled {
        border-color: var(--s-color-base-content-quaternary);
      }
      &:active {
        background-color: var(--s-color-base-disabled);
      }
      &:disabled,
      &:disabled:hover {
        background-color: transparent;
        color: var(--s-color-base-on-disabled);
      }
    }
  }
  @include full-width-button;
  @include full-width-button('el-button--create-pair', $inner-spacing-mini);
}

.pool {
  &-list {
    width: 100%;
    border-top: none;
    border-bottom: none;
    .pool-info-container {
      margin-bottom: $inner-spacing-mini;
      padding: calc(#{$inner-spacing-big} - (#{$pair-icon-height} - var(--s-size-small)) / 2) $inner-spacing-medium;
      &:last-child {
        margin-bottom: 0;
      }
    }
    h3 {
      letter-spacing: $s-letter-spacing-small;
      font-feature-settings: $s-font-feature-settings-title;
    }
  }
  &-info {
    display: flex;
    align-items: center;
    font-size: var(--s-font-size-small);
    line-height: $s-line-height-big;
    margin-bottom: $inner-spacing-mini;
    &,
    &--buttons {
      padding-left: $inner-spacing-mini;
      padding-right: $inner-spacing-mini;
    }
    &--share {
      color: var(--s-color-base-content-secondary);
    }
    &-value {
      margin-left: auto;
    }
    &-container {
      width: 100%;
      padding: $inner-spacing-big;
      border-radius: var(--s-border-radius-small);
      border: 1px solid var(--s-color-base-border-secondary);
      color: var(--s-color-base-content-tertiary);
      font-size: var(--s-font-size-mini);
      line-height: $s-line-height-small;
      font-feature-settings: $s-font-feature-settings-common;
      text-align: center;
      & + .el-button {
        margin-top: $inner-spacing-medium;
      }
    }
    & > .token-logo {
      margin-right: $inner-spacing-mini;
    }
    &--buttons {
      display: flex;
      margin-top: $inner-spacing-medium;
      .el-button {
        padding-left: $inner-spacing-small;
        padding-right: $inner-spacing-small;
        font-feature-settings: $s-font-feature-settings-title;
        width: auto;
        @include font-weight(700);
        + .el-button {
          margin-left: $inner-spacing-mini;
        }
      }
    }
  }
}
</style>
