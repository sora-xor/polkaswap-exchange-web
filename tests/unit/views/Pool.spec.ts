import { shallowMount, createLocalVue } from '@vue/test-utils'

import Pool from '@/components/Pool.vue'
import { TranslationMock } from '../../utils'

const localVue = createLocalVue()

describe('Pool.vue', () => {
  beforeEach(() => {
    TranslationMock(Pool)
  })

  it('should renders correctly', () => {
    const wrapper = shallowMount(Pool, { localVue })
    expect(wrapper.element).toMatchSnapshot()
  })
})
