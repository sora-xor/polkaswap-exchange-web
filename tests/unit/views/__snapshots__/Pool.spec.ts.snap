// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pool.vue should renders correctly 1`] = `
<s-form>
  <s-form-item>
    Pool content
  </s-form-item>
</s-form>
`;
